New Yaml (and JSON) file/shell collector; some rando Makefile fixes.

mikepea opened the pull request:
This collector gets YAML (or JSON) from a source and returns metrics for each
k/v found at the top level that can be converted to a float (including
booleans).

Sources can be a file (eg /etc/mcollective/facts.yaml), or a shell pipeline or simple exec.

This is designed to be run with multiple instantiations, for different sources and intervals.

## yamlMetricsKeyWhitelist option in YamlMetrics collector

When reading arbitrary YAML files - like the facter/mco /etc/mcollective/facts.yaml file -
it's necessary to whitelist certain keys, to prevent processing what could be
quite a lot of data and creating unnecessary/confusing signalfx DPs.

## Boolean support to YamlMetrics

stringy or proper boolean values in Simple (eg Factor-like) Yaml are converted to 1 (true) and
0 (false).

## Fix up glide.* to import yaml libs; fix annoying case issue with logrus

Having two different case-versions of sirupsen/logrus (vs Sirupsen/logrus)
was really freaking out glide. I could not get it to bring in the new yaml libs
or generally play nice.

## Add JSON tests, as we get it for free with YAML!

As JSON is a subset of YAML, we can read JSON from exec/file. Highlight this is possible in the tests.

## Add 'exec' and 'shell' sources for YAML.

The file-based YAML is only so useful, and requires that you have a secondary process writing to a file (maybe puppet, or a cron job like `factor --yaml` for mCollective). Fullerite is already a pretty good interval-based scheduler, so let's allow it to run arbitrary commands to produce the YAML/JSON. 

Prefix a yamlSource with 'exec:' to cause the command to be run directly.

Prefix a yamlSource with 'shell:' to run it under 'sh -c', so that pipelines
and args are functional. Eg `http JSON_ENDPOINT | jq 'stuff'` or `myscript arg1 arg2`.

Commands must output valid YAML (or JSON), in either Simple or Fullerite format.

## 'Fullerite' format

It's not really possible with the 'read arbitrary YAML/JSON' idea to have per metric dimensions. This is quite limiting for many use-cases, and so it's advantageous to have a format that allows them.

By specifying 'format: "fullerite"' as a top-level key, read metrics.Metrics format metric+dimensions+value structs from an list under 'metrics: []' key.

dichiarafrancesco commented on src/fullerite/collector/yaml.go:
```diff
@@ -0,0 +1,248 @@
+	return y, err
+}
+
+func extractYamlSourceMethodAndSource(s string) (string, string) {
```
why not having two different configs for sourceMethod and source? doesn't make things easier?

dichiarafrancesco replied:
not a blocking issue.

dichiarafrancesco commented on src/fullerite/collector/yaml.go:
```diff
@@ -0,0 +1,248 @@
+	err := yaml.Unmarshal(yamlData, &m)
+	if err != nil {
+		c.log.Error("Could not unmarshal YAML: ", err.Error())
+		c.log.Debug(fmt.Sprintf("%s", yamlData))
```
maybe you forgot **return metrics**??

mikepea replied:
good point, fixing.

dichiarafrancesco commented on src/fullerite/collector/yaml.go:
```diff
@@ -0,0 +1,248 @@
+// simple k/v format; only keep values that convert to float
+func (c *YamlMetrics) getSimpleFormatMetrics(m map[string]interface{}) (metrics []metric.Metric) {
+	for k, v := range m {
+		if len(c.yamlKeyWhitelist) > 0 && !c.yamlKeyMatchesWhitelist(k) {
```
as we said yesterday, maybe it's safer to check the whitelist all the time. so if the whitelist is empty we don't send anything (instead of sending everything). the risk with the current code is that it might happen to generate lots of new keys at once, causing problems on signalfx side. Also it'll prevent to go above our DPM threshold.

dichiarafrancesco replied:
what about this? I think it's pretty important

mikepea replied:
agreed, sorry missed it in the PR update. Fixing now.

dichiarafrancesco commented on src/fullerite/collector/yaml.go:
```diff
@@ -0,0 +1,248 @@
+func (c *YamlMetrics) buildMetric(k string, v float64) metric.Metric {
+	metricName := k
+	if c.metricPrefix != "" {
+		metricName = c.metricPrefix + "." + k
```
I think this should be mandatory for what said before.

mikepea commented on src/fullerite/collector/yaml.go:
```diff
@@ -0,0 +1,248 @@
+)
+
+const (
+	defaultYamlCollectorName = "YamlMetrics"
```
I don't follow. This is the 'YamlMetrics collector'.

I can't see 'Collector' in any collector names in src/fullerite/collectors.

mikepea commented on src/fullerite/collector/yaml.go:
```diff
@@ -0,0 +1,248 @@
+//     (via remarshalling to JSON and using standard JSON unmarshal into object)
+//
+//  Config:
+//     metricPrefix     - prefix to add to Metrics, default 'yamlMetrics'. "" for no prefix.
```
Happy to make it mandatory, better safe than sorry.

mikepea commented:
@dichiarafrancesco sorry for the delay, been unable to spend time working on this recently. I've updated as per your needs, and also make 'simple' mode less of a risk (both in code as discussed but also by removing it as the default).

I've simplified 'fullerite' mode to be compatible with the adhoc collector, which could be useful.

dichiarafrancesco commented:
Neat. merging it :)
